### ISSUE TITLE
Memory Saving Gradients for TF2

### ISSUE THREAD

**Issue by mpafla** [tensorflow/tensorflow#30418]
**System information**
- TensorFlow 2.0 beta

There are libraries for TF1 that are able to calculate more memory efficient gradients such as [gradient-checkpointing](https://github.com/cybertronai/gradient-checkpointing). They edit the graph with the [tf.contrib.graph_editor](https://www.tensorflow.org/api_docs/python/tf/contrib/graph_editor) to save memory. However, tf.contrib does not exist anymore which is why I cannot run large models such as GPT2 in TF2 (converted with `tf_upgrade_v2`) as it overflows my GPU memory in Google Colab. Is there a TF2 feature that can help me?

**Comment by gadagashwini** [comment]
@mpafla Will it be possible to provide us the minimal code snippet which replicates your issue.Thanks!

**Comment by ppwwyyxx** [comment]
This issue is a feature request, so what's there to replicate?

**Comment by mpafla** [comment]
If you have a look [here](https://github.com/nshepperd/gpt-2/blob/finetuning/src/memory_saving_gradients.py), they use `import tensorflow.contrib.graph_editor as ge` which does not exist in TF2. How would I do this in TF2? Is there some feature that helps me with this?

Sorry, I cannot provide a code snippet. I need some general guidance (maybe I should've posted on stackoverflow - I wasn't sure).

**Comment by alextp** [comment]
There's the [tf.recompute_grad](https://github.com/tensorflow/tensorflow/blob/2cb8b4ab8f038f5c6de2381ad215fdaa6fd2bc22/tensorflow/python/ops/custom_gradient.py#L351) decorator available in tf2 nightly which you can add to your code to do checkpointing.

It's probably possible to fork tf.contrib.graph_editor and make it work with tf.compat.v1.wrap_function to keep the functionality you need, though.

Otherwise I'm happy to discuss proposals for a better tf2-friendly way of making this happen.

**Comment by mpafla** [comment]
Thanks for your answer, I think it could be done with this. I am far from an expert, but I like how it is done [here](https://github.com/cybertronai/gradient-checkpointing/blob/master/memory_saving_gradients.py#L42). Basically, you overwrite tf.gradients with their version of gradients, but are also to specify a 'checkpoint heuristic' such as 'memory' or 'speed' that computes gradients more memory or more speed efficient respectively.

**Comment by davisyoshida** [comment]
@alextp I'm not sure that recompute_grad allows for gradient checkpointing style memory reduction.

The code below runs 200 blocks of 200 "layers", each of which is just element-wise operations with a single vector. Gradient checkpointing should mean that during the forward pass, the activations of only 200 layers are stored, rather than all 40000 layers.

If this were accomplishing what gradient checkpointing does, the peak memory usage should be around 100000 * 200 * 2 * 4 bytes (<200 MB), but this causes an OOM error on my 8 GB GPU. Have I understood the suggestion correctly?

```
import numpy as np
import tensorflow as tf

DIM = 100000
B_SIZE = 200
@tf.recompute_grad
def run_block(x, variables=None):
    for _ in range(B_SIZE):
        x = variables[0] * x
    return x

v = tf.Variable(np.zeros(DIM), dtype=tf.float32)
x = tf.zeros(DIM, dtype=tf.float32)
opt = tf.optimizers.SGD(lr=1e-5)

with tf.GradientTape() as g:
    g.watch(x)
    layer_out = x
    for _ in range(B_SIZE):
        layer_out = run_block(layer_out, variables=[v])
    loss = tf.reduce_sum(layer_out)
    print(g.gradient(loss, x))
```

**Comment by davisyoshida** [comment]
@mpafla I've written a simple gradient checkpointing decorator [here](https://github.com/davisyoshida/tf2-gradient-checkpointing)

**Comment by ppham27** [comment]
I believe the tf-slim version works:

https://github.com/google-research/tf-slim/blob/a62dc893de5e46e6f2e9ec24a74b2abce026307a/tf_slim/layers/rev_block_lib.py#L465

**Comment by davisyoshida** [comment]
It doesn't appear that there's a way to use that for something like a Keras layer right?

**Comment by ppham27** [comment]
It should work fine inside a Keras layer, something like:

```python
class MyLayer(tf.keras.layers.Layer):

  def call(self, inputs):
    @rev_block_lib.recompute_grad
    def call_impl(inputs):
      # do stuff

    return call_impl(inputs)
```

**Comment by davisyoshida** [comment]
Does it only work in graph mode perhaps? I tried it in eager mode and didn't get memory savings.

**Comment by mathemakitten** [comment]
Hi @ppham27, just chiming in to +1 definitely interested in an implementation of gradient checkpointing which works in graph mode. The tf-slim version doesn't run for TF2+ versions which no longer have contrib (without quite a bit of hackery, anyway).

**Comment by ppham27** [comment]
> Does it only work in graph mode perhaps? I tried it in eager mode and didn't get memory savings.

I'm not sure if it works in pure eager mode, but I do recall it working with TF2 if you annotated your computation with `@tf.function` (possibly with `experimental_compile=True`). This was months ago since I last tried, and perhaps, the situation is different, now.

> Hi @ppham27, just chiming in to +1 definitely interested in an implementation of gradient checkpointing which works in graph mode. The tf-slim version doesn't run for TF2+ versions which no longer have contrib (without quite a bit of hackery, anyway).

Okay, I'll see what it takes to push something to https://github.com/tensorflow/addons. I'm not on the TF team at Google and only contribute in my spare cycles, so it may take me some time. My team's implementation is tested to work with both TF1 and TF2, but there still may be some cases, we've missed.

**Comment by mathemakitten** [comment]
Just updating that I spent some time extricating the necessary code out of tf-slim and it runs with `tf-nightly==2.2.0.dev20200303` in graph mode successfully for a model running on TPU.

**Comment by RayXie29** [comment]
@mathemakitten Do you mind to share how you did this on TPU? I can only manage to make it work on GPU model training, by splitting the model into several blocks and apply recompute_grad on them. But in TPU, the model seems won't work if I split the model into several blocks for training.

**Comment by mathemakitten** [comment]
@RayXie29 I've uploaded a standalone version of the code I extracted out of the tf-slim implementation for usage on TPU and made it available [here](https://github.com/mathemakitten/gradient-checkpointing). I hope it helps!

**Comment by RayXie29** [comment]
@mathemakitten Thank you so much for sharing this.

**Comment by ppham27** [comment]
If anybody still needs this, we released [recompute_grad](https://github.com/google-research/google-research/blob/master/etcmodel/layers/recompute_grad.py#L169) as part of our EMNLP paper [ETC: Encoding Long and Structured Inputs in Transformers
](https://arxiv.org/abs/2004.08483)

We ran our experiments on TPUs in the paper, but I've tested on GPUs with and without XLA. It has the nice bonus that it supports dropout layers, too, if you use [RecomputingDropout](https://github.com/google-research/google-research/blob/81200ee32d8ce89e2359cd2bc4e3bd29ad713ba6/etcmodel/layers/recomputing_dropout.py#L61), which should be a drop-in replacement for `tf.keras.layers.Dropout`.

**Comment by rmothukuru** [comment]
@mpafla,
With respect to the [above comment](https://github.com/tensorflow/tensorflow/issues/30418#issuecomment-719124842), can you please let us know if we can close this issue? Thanks!

**Comment by sanatmpa1** [comment]
@mpafla,

As I can see that you've reacted thumbs up to @rmothukuru 's above comment about closing the issue, I am closing it now. Thanks!

**Comment by nyngwang** [comment]
@ppham27 Thanks for your sharing! Just some checks:

1. Does it work with tf v2.7.0 (installed from conda-forge channel)?
2. Does it work with both eager mode and graph mode?
3. (optional) I'm relatively new to this. Could you point me to some sample code as an entry point for learning how to use it?

Thanks for your reading.

**Comment by ppham27** [comment]
1. We run unit tests against the latest version of TF, so yes.
2. It is different, since it has special support for XLA:GPU and XLA:TPU
3. it should work in both eager and graph mode. Inside of eager mode, it only works when wrapped with tf.function
4. Yes, it's tested on GPUs.

Also, it is only tested with `tf.function(jit_compile=True)`. I am unsure if it works in other cases. I think i was told it worked with just `tf.function` but I never verified it myself.

**Comment by nyngwang** [comment]
@ppham27 I'm really happy to see your kindly, to-the-point reply! Due to some reasons, I was told to re-write the Repo. https://github.com/cybertronai/gradient-checkpointing to TensorFlow 2 version. This made me sad because I'm really new to Tensorflow v2. So I need to ask you some more questions:

1. Does your implementation already achieve what I was told to do: re-rewriting the (original?) Repo. of gradient-checkpointing? If so, would you mind me asking questions about both your Repo. and the linked Repo., if I have any questions regarding your implementation? (this could probably change my life, seriously)
2. If 1. is hard to answer, then from your point of view, for one who is a very beginner of TensorFlow (I'm OK with python), how much time might be required to achieve 1., from your experience.

Thanks for your reading...

**Comment by ppham27** [comment]
I am not too familiar with that repository, unfortunately. I don't see a lot of value in rewriting it, though. If I were to keep pursuing this, I'd be more inclined to merge this into TF Core as a fix for https://www.tensorflow.org/api_docs/python/tf/recompute_grad.

**Comment by nyngwang** [comment]
@ppham27 I just updated my last comment I was writing it in a hurry.

> I don't see a lot of value in rewriting it, though.

It seems that it depends on some modules that have been removed in TF2, so I was asked to do so.

**Comment by nyngwang** [comment]
@ppham27 Thanks for your clarification!

**Comment by nyngwang** [comment]
Does your implementation provide `__wrapped__` attribute like the one [`tf.recompute_grad`](https://www.tensorflow.org/api_docs/python/tf/recompute_grad)?

**Comment by ppham27** [comment]
Hi @nyngwang,

No it does not, but it's trivial to add if you want to fork it: https://github.com/tensorflow/tensorflow/blob/57e779243d459bd2c3c7c2b1d54ed1a1adfa56b6/tensorflow/python/util/tf_decorator.py#L100

**Comment by Apprisco** [comment]
@ppham27 Hello! I've been trying to use your gradient checkpointing solution in tf 2.9.1 with jit compilation enabled, but keep running into an OOM that I do not face if I don't enable gradient checkpointing. I'm using a double stacked model: as in, I have a VAE with a separate encoder and decoder model. Do you have any easier-to-read docs for your gradient checkpointing py, and have you verified that it works on 2.9.1?

**Comment by Apprisco** [comment]
Update: I was able to get it working on tensorflow 2.9.1, w/o jit_compile= True

**Comment by Apprisco** [comment]
Update v2: turns out enabling gradient checkpoints in your manner will prevent model from training- it seems watch vars is simply not transferring.

**Comment by nyngwang** [comment]
@ppham27 Could you provide a minimal example of how to use it in graph mode? I have the same result as @Apprisco.

**Comment by ppham27** [comment]
Sorry for the trouble. I have pretty limited time these days, but if you have a repro in colab, I might be able to play with it over the next couple of weeks.
